packages/connector-puppeteer/src/connector.ts

```diff
@@ -0,0 +1,381 @@
+        if (!isRegularProtocol(target.href)) {
+            const error = new Error(`Target protocol is not valid (is ${target.protocol})`);
+
+            (error as any).type = 'InvalidTarget';
+
+            throw error;
+        }
+
+        const { browser, page } = await launch(this._options);
+
+        this._browser = browser;
+        this._page = page;
+    }
+
+    private async processTarget() {
+        await this.waitForTarget();
+
+        const event = { resource: this._finalHref };
+
+        // QUESTION: Even if the content is blank we will receive a minimum HTML with this. Are we OK with the behavior?
+
+        const html = await this._page.content();
+
+        this._dom = createHTMLDocument(html, this._originalDocument);
+
+        // Process pending requests now that the dom is ready
+        while (this._pendingRequests.length > 0) {
+            const pendingRequest = this._pendingRequests.shift()!;
+
+            await pendingRequest();
+        }
+
+        if (this._options.headless) {
+            // TODO: Check if browser downloads favicon even if there's no content
+            await getFavicon(this._finalHref, this._dom, this.fetchContent.bind(this), this._engine);
+        }
+
+        if (this._targetBody) {
+            await traverse(this._dom, this._engine, this._page.url());
+
+            await this._engine.emitAsync('can-evaluate::script', event);
+        }
+    }
+
+    public async close() {
+        this.removeListeners();
+
+        await close(this._browser, this._page);
+    }
+
+    public evaluate(code: string): Promise<any> {
+        return this._page.evaluate(code);
+    }
+
+    public async collect(target: URL): Promise<void> {
+        await this.initiate(target);
+
+        await this._engine.emit('scan::start', { resource: target.href });
+
+        // TODO: Figure out how to execute the user tasks in here and when to subscribe to events
+
+        this.addListeners();
+
+        debug(`Navigating to ${target.href}`);
+        await this._page.goto(target.href, { waitUntil: this._waitUntil });
+        // TODO: what happens if there are multiple redirects here?
+
+        // This is the final URL
+        this._finalHref = this._page.url();
+
+        debug(`Navigation complete`);
+
+        // Stop network events after navigation is considered complete so no more issues are reported
+        this.removeListeners(['request', 'requestfailed', 'response']);
+
+        await this.processTarget();
+
+        // Some other timeouts or awaits here?
+
+        await this._engine.emitAsync('scan::end', { resource: this._finalHref });
+    }
+
+    public fetchContent(target: URL | string, customHeaders?: object): Promise<NetworkData> {
+        /*
+         * TODO: This should create a new tab, navigate to the
+         * resource and control what is received somehow via an event.
+         */
+        const assigns = compact([this && this._headers, customHeaders]);
```

Comment:
After digging into this a bit, I'm wondering if `this._headers` can/should be dropped entirely, both here and in the class as a whole. These are headers associated with the `target` request, but we're potentially applying them to requests for other resources as well which seems wrong. Plus most re-requests already pass the headers from the original request as part of `customHeaders`.